File: C/Timer.c
#include <stdbool.h>
#include "./include/Timer.h"

void timer_update(struct CHIP8* self) {
    static bool sound_is_playing = false;
    //static tone

    if (self->DT > 0) {
        self->DT -= 1;
    }

    if ((self->ST > 0) && (sound_is_playing == false)) {
        self->ST -= 1;
        // tone play
        sound_is_playing = true;
    }
    else {
        if ((self->ST > 0) && (sound_is_playing == true)) {
            self->ST -= 1;
        }
        else {
            if ((self->ST <= 0) && (sound_is_playing == true)) {
                // tone fadeout
                sound_is_playing = false;
            }
        }
    }
}

File: Rust/src/timer.rs
use super::CHIP8;

const DT: usize = 0;
const ST: usize = 1;

pub fn timer_update(interpreter: &mut CHIP8) {
    
    static mut SOUND_IS_PLAYING: bool = false;
    // static tone
    
    if interpreter.timers[DT] > 0 {
        interpreter.timers[DT] -= 1;
    }

    unsafe {
        if (interpreter.timers[ST] > 0) && !SOUND_IS_PLAYING {
            interpreter.timers[ST] -= 1;
            // tone play
            SOUND_IS_PLAYING = true
        } else if (interpreter.timers[ST] > 0) && SOUND_IS_PLAYING {
            interpreter.timers[ST] -= 1;
        } else if interpreter.timers[ST] <= 0 {
            // tone fadeout
            SOUND_IS_PLAYING = false;
        }
    }
}